Release GIL when doing standalone solves

jberg5 opened the pull request:
Addresses https://github.com/Simple-Robotics/proxsuite/issues/358.

Releases the Global Interpreter Lock (GIL) when using `proxsuite.proxqp.dense.solve` so python users can easily parallelize solving problems across cores. Quadprog and OSQP already do something similar.

This enables 2x or more faster parallel solving compared to using `proxsuite.proxqp.dense.VectorQP()` and `solve_in_parallel`.

On my M2 Mac, I see the following:
```
python ../benchmark/timings-parallel.py
Problem specs: n=1000 n_eq=200 n_in=200. Generating 64 such problems.
Generated problems. Solving 64 problems with 11 threads.
...
setup_vector_of_qps: 4655.346ms
setup_vector_and_solve_parallel_heuristics_threads: 5850.690ms
solve_serial_dense: 4716.554ms
solve_parallel_dense: 1934.598ms
```

Headline result is that solving the problems using the standalone `dense.solve` and a `ThreadPoolExecutor` is more than 2x faster than solving the problems serially now that the GIL gets released, and almost 3x faster than setting up a `VectorQP` and running `solve_in_parallel` (iiuc the latter mostly being a result of the problem conditioning happening when building the vector, which is a serial operation). If the GIL weren't released, then using a `ThreadPoolExecutor` would only add overhead and would take longer than just solving the problems serially.

As an additional benefit, this change doesn't require building proxsuite with OpenMP, which I think is what has kept `solve_in_parallel` from being released on all platforms.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Simple-Robotics/proxsuite?pullRequest=359) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Simple-Robotics/proxsuite?pullRequest=359) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Simple-Robotics/proxsuite?pullRequest=359) it.</sub>

jberg5 commented:
Hi @fabinsch, thank you for your comment! I would be happy for you to apply your changes to the benchmark script on top of my branch. Once you do I'll re-run on my machine and put some results here.

Your results with n=50 and n=500 are really interesting. I can confirm that I get similar results running the same problem configuration on my machine. I confess I had only run tests on a scale that was closer to my use case (n >= 1000) and I don't have a good intuition for why n=500, n_eq = n_in = 200 is so slow in a threadpool. Might be interesting to run this benchmark against a lot of different problem setups and plot the results.

Re: `proxsuite.proxqp.dense.solve_no_gil`, as long as the implementation of `proxsuite.proxqp.dense.solve` is already threadsafe, then I don't think we would need a separate `solve_no_gil`. Releasing the GIL wouldn't impact thread safety - it just means that the Python program can do other things while the c++ crunches the numbers. I can update the changelog and the docs to make the GIL change explicit just so users know that they can take advantage of the new behavior. Let me know if this addresses your concerns!

fabinsch commented:
Thank you for the answer, I pushed my changes.

One concern I have is that our implementation uses [MatRef](https://github.com/Simple-Robotics/proxsuite/blob/main/bindings/python/src/expose-solve.hpp#L23) (an [Eigen::Ref](https://github.com/Simple-Robotics/proxsuite/blob/main/include/proxsuite/proxqp/dense/fwd.hpp#L33) to the input data), which directly references the underlying NumPy arrays without copying them. If the GIL is released during the execution of `dense.solve` and the user modifies the input arrays (e.g., resizing or changing elements) in Python, this could lead to undefined behavior.

Existing users of `dense.solve` might rely on it holding the GIL for specific use cases and changing the behavior could unintentionally break their workflows. That's why I think it should be intentional to release it, by either using a new function or passing an additional keyword arg like `release_gil=True`.

What do you think about it @jorisv or @jcarpent ?

jorisv commented:
@fabinsch  I agree with you. It's better to add this as an option. So advanced user can release the GIL, but standard user will benefit of this safety.

jcarpent commented:
> @fabinsch I agree with you. It's better to add this as an option. So advanced user can release the GIL, but standard user will benefit of this safety.

I do agree too.

jberg5 commented:
@fabinsch @jorisv @jcarpent thanks guys, appreciate the feedback! Separate API sounds good. Just updated my branch - lmk if it looks good to you.

Regarding timings, I got very similar results on my machine (had to reduce the number of problems from 128 to 32 to get everything to fit in one process on my machine), where the n=500 case is slower with the ThreadPoolExecutor than even just solving everything serially, which is a very interesting result. I'll investigate some more. All the other test cases (50, 100, 200, 1000) were faster with the _no_gil methods and ThreadPoolExecutor.

fabinsch commented:
This PR here looked good to me, all tests pass locally. I wanted to run our ci, which failed with an unrelated miniforge error when trying to create the conda env. When I tried to push a commit on top, I accidentally merged it.

I had to reopen a new PR (#363). I apply exactly the same commits as here, plus I try to fix the conda ci to check if all is good.
